### PR TITLE
8240692: Cleanup of the javafx property objects

### PR DESCRIPTION
**Pull request by nlisker**
A simple readability cleanup for the files that were changed in #113.

Note that the boolean property has an `instanceof` check that doesn't exist in the other properties.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8240692](https://bugs.openjdk.java.net/browse/JDK-8240692): Cleanup of the javafx property objects

### Reviewers
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/141/head:pull/141`
`$ git checkout pull/141`

### CODE REVIEW

**Comment by kevinrushforth** [comment]
I note that this also changes the wrapper property objects from anonymous subclasses of XxxxxPropertyBase to SimpleXxxxxProperty. This is more than just a readability cleanup. It's probably fine for this case, but that's why I want a second reviewer.

**Comment by nlisker** [comment]
> I note that this also changes the wrapper property objects from anonymous subclasses of XxxxxPropertyBase to SimpleXxxxxProperty. This is more than just a readability cleanup. It's probably fine for this case, but that's why I want a second reviewer.

Isn't SimpleXxxxxProperty exactly made for XxxxxPropertyBase with the built-in overrides for the bean and the name? When is this substitution not fine?

**Comment by tomsontom** [comment]
I can somehow remember asking Richard Bair why JavaFX internally does not use Simple* but creates the anonymous subclasses and he said it's memory reason - Simple* uses more memory because of the additional fields

**Comment by nlisker** [comment]
That doesn't seem right. The additional fields are captured in the
anonymous class anyway (same as in lambdas).

**Comment by tomsontom** [comment]
the subclass saves the owner field who is a static null, not?

**Comment by nlisker** [comment]
From what I see, they save the same data.

**Comment by kevinrushforth** [comment]
The anonymous subclasses don't have an instance field for the `bean` or the `name`, so the tradeoff is one of static footprint (an extra anonymous inner class) for dynamic footprint (slightly larger instances).

Note, however, that this savings is reduced by the fact that the inner classes have a `this$0` pointer to the enclosing class. I think this means that we are really only saving one reference variable per instance of the anonymous inner class versus an instance of SimpleXxxxProperty. That's why I think that in the specific case of this proposed fix, it should be fine. I just want to take a closer look.

**Comment by kevinrushforth** [comment]
I ran tests to validate my above assumption, and the results match my expectations. Using a `SimpleXxxxxProperty` takes an extra 8 bytes (one object reference) of storage versus an anonymous inner class. I used a null bean and a `this$0` reference to get the name. I measured it by checking the memory needed for 1,000,000 instances of an anonymous subclass of `DoublePropertyBase` objects versus 1,000,000 instances of `SimpleDoubleProperty`. The former takes 40 bytes per instance while the latter takes 48 bytes. This represents a 20% increase in memory used.

To see how much this affects the case of the wrapped property objects, I ran two tests on the current code and the same two tests with your proposed fix. The first test creates an `ObjectProperty<Double>` object from a `DoubleProperty` using `DoubleProperty::asObject`. The second test created a `DoubleProperty` object from an `ObjectProperty<Double>` using `DoubleProperty::doubleProperty`.

I got the same results as for the earlier tests: 8 additional bytes (one object reference) using `SimpleXxxxProperty` to create the new wrapped property object. Since these objects are already fairly large due to the bidirectional binding created between the original property and the wrapping property, the percentage increase is not all that great -- about 3.8% in the case of `xxxxProperty`.

| Method | bytes/object current | bytes/object with patch |
| --- | --- | --- |
| DoubleProperty::asObject | 232 | 240 |
| DoubleProperty::doubleProperty | 208 | 216 |

So this begs the question: Do we still want to do this? Is the cleanup worth the extra memory used?

**Comment by nlisker** [comment]
> So this begs the question: Do we still want to do this? Is the cleanup worth the extra memory used?

I'm not sure. This raises the opposite question: are there any places where we can benefit from introducing anonymous classes instead of the current implementation?

In my own code, I prefer the readability and conciseness over performance. However, since JavaFX is a library, we might want to go with the performance route. I don't know how important the memory consumption is for users in this case or the other similar cases.

**Comment by kleopatra** [comment]
Just playing devil's advocate :)

Technically, there's a functional difference between the inline extension and useage of SimpleXX: in the former, getName returns the current name of mapped property, in the latter it returns its name at the time of creating the mapped property. 

Afaics, there is nothing in the spec of getName that prevents custom implementations with name (and bean) being mutable (the SimpleXX have implemented them immutable, but then they are only default implementations). Not that I can imagine any use-case for it, but if they exist, the new mappings might break existing code.

**Comment by kevinrushforth** [comment]
Getting back to this, I think we shouldn't make the change from inner classes to SimpleXxxxProperty objects proposed by this PR.

As for whether we should consider changing some internal instance of SimpleXxxxProperty (which I think wouldn't run afoul of the hypothetical problem raised by @kleopatra ), we could file a new Enhancement request if you like, but it's probably not all that critical, unless we were to find a few in the base Node or Control classes that could yield big savings.

File: modules/javafx.base/src/main/java/javafx/beans/property/BooleanProperty.java [hunk 1]
```diff
@@ -180,4 +180,4 @@ public String getName() {
             }
         };
     }
-}
+}
```
Comment (kevinrushforth) [inline modules/javafx.base/src/main/java/javafx/beans/property/BooleanProperty.java]: Minor: can you add the trailing newline back in? We typically end each file with a single new-line; also this shows up as an unnecessary diff.

**Comment by kevinrushforth** [comment]
Given that this PR is now just a cleanup fix, it is simple enough that it doesn't need a second reviewer.
